Add fix function for #919

joaquimg opened the pull request:
fix candidate for #919

mlubin commented on src/JuMP.jl:
```diff
@@ -391,6 +391,14 @@ function setvalue(v::Variable, val::Number)
 
+# Fix a variable that was not previously fixed
+function fix(v::Variable, val::Number)
+    v.m.colVal[v.col] = val
```
Could ``setvalue`` and ``setcategory`` instead of the internal methods?

mlubin commented:
Needs tests and docs.

joaquimg commented:
This one runs smoothly:
```function fix(v::Variable, val::Number)
    v.m.colCat[v.col] = :Fixed
    setvalue(v, val)
end```
While this one does not work:
```function fix(v::Variable, val::Number)
    setcategory(v, :Fixed)
    setvalue(v, val)
end```
Due to:
```
const var_cats = [:Cont, :Int, :Bin, :SemiCont, :SemiInt]
function setcategory(v::Variable, cat::Symbol)
    cat in var_cats || error("Unrecognized variable category $cat. Should be one of:\n    $var_cats")
    v.m.colCat[v.col] = cat
end
```
Clearly ``:Fixed`` is not allowed...

mlubin commented:
Okay, you could argue that using ``setcategory`` is not appropriate for fixed variables because you should also update the bounds at the same time to maintain consistency. I wonder if we should remove the special behavior from ``setvalue`` and just use ``fix`` to update the values of fixed variables.

joaquimg commented:
I think the current behavior is ambiguous and makes a little confusion, although I added the docs.

I would prefer ``setvalue`` only for warmstarts and ``fix`` for fixed values (allowing the user to fix previously non-fixed variables)

joaquimg commented:
Lets go for  

> remove the special behavior from `setvalue` and just use `fix` to update the values of fixed variables.

?

mlubin commented:
Yes. Calling ``setvalue`` on a fixed variable should (for now) throw an error since this is a breaking change and we don't want code silently changing behavior.

mlubin commented on src/JuMP.jl:
```diff
@@ -386,15 +386,16 @@ getupperbound(v::Variable) = v.m.colUpper[v.col]
     if v.m.colCat[v.col] == :Fixed
-        v.m.colLower[v.col] = val
-        v.m.colUpper[v.col] = val
+        error("setvalue use for setting bounds of fixed variables was deprecated in favour of JuMP.fix(v::Variable, val::Number), which changes bounds of fixed variables and fixes previously not fixed variables ")
```
Deprecated would mean that we still support it but throw a warning. This isn't the case. ``s/favour/favor/``

joaquimg replied:
what do I say?
`setvalue use for setting bounds of fixed variables was made obsolete in favor of JuMP.fix(v::Variable, val::Number), which changes bounds of fixed variables and fixes previously not fixed variables`

mlubin replied:
``setvalue for fixed variables is no longer supported. Use JuMP.fix instead.``

mlubin commented on doc/refvariable.rst:
```diff
@@ -219,7 +219,9 @@ Two possible uses for fixed variables are:
    in particular, most solvers are able to efficiently hot-start when
    solving the resulting modified problem.
+   
+Note that ``setvalue(x, val)`` will not set initial value for fixed variables, it will only set a warmstart for non-fixed variables. One can fix existing, non-fixed variables with ``JuMP.fix(x, val)`` which sets both bounds and initial value of ``x`` to ``val``.
```
I don't think this paragraph is needed

mlubin commented on doc/refvariable.rst:
```diff
@@ -219,7 +219,9 @@ Two possible uses for fixed variables are:
 
 2. For solving a sequence of problems with varying parameters.
-   One may call ``setvalue(x, val)``
+   One may call ``JuMP.fix(x, val)``
```
One may call ``JuMP.fix(x,val)`` to change the value the which the variable is fixed or to fix a previously unfixed variable.

mlubin commented:
Almost ready to merge, but it's currently failing tests:
```
Test Summary:                                 | Pass  Error  Total
  Variables                                   |   85      1     86
  constructors                                |   11            11
  get and set bounds                          |   11      1     12
```

mlubin commented:
There are still failing tests

joaquimg commented:
I will look into that!

joaquimg commented:
I had no errors in Variables, also I could not notice errors in Xpress and GLPK

mlubin commented:
See https://travis-ci.org/JuliaOpt/JuMP.jl/jobs/190050309#L409

joaquimg commented:
Now its passing all tests!
